src/formElementPlugins/listBuilder/src/helpers.ts

```diff
@@ -0,0 +1,112 @@
+import { useUserState } from '../../../contexts/UserState'
+import { DisplayType, InputResponseField, ListItem, ListLayoutProps } from './types'
+
+// Formatting and Text manipulation
+export const getDefaultDisplayFormat = (inputFields: TemplateElement[]) => {
+  const displayString = inputFields.reduce(
+    (acc: string, { code, title }) => acc + `**${title}**: \${${code}}  \n`,
+    ''
+  )
+  return { title: '', subtitle: '', description: displayString }
+}
+
+export const substituteValues = (parameterisedString: string, item: ListItem) => {
+  const getValueFromCode = (_: string, $: string, code: string) => item[code]?.value?.text || ''
+  return parameterisedString.replace(/(\${)(.*?)(})/gm, getValueFromCode)
+}
+
+export const createTextString = (listItems: ListItem[], inputFields: TemplateElement[]) =>
+  listItems.reduce(
+    (outputAcc, item) =>
+      outputAcc +
+      inputFields.reduce(
+        (innerAcc, field) => innerAcc + `${field.title}: ${item[field.code]?.value?.text}, `,
+        ''
+      ) +
+      '\n',
+    ''
+  )
+
+// Data validation and state reset
+export const resetCurrentResponses = (inputFields: TemplateElement[]) =>
+  inputFields.reduce((acc, { code }) => ({ ...acc, [code]: { value: { text: undefined } } }), {})
+
+export const anyInvalidItems = (currentInput: ListItem) =>
+  Object.values(currentInput).some((response) => response.isValid === false)
+
+export const anyIncompleteItems = (currentInput: ListItem, inputFields: TemplateElement[]) =>
+  Object.values(currentInput).some(
+    (response, index) => inputFields[index]?.isRequired !== false && !response.value?.text
+  )
+
+export const anyErrorItems = (currentInput: ListItem, inputFields: TemplateElement[]) =>
+  anyInvalidItems(currentInput) || anyIncompleteItems(currentInput, inputFields)
+
+// Building elements
+export const combineResponses = (
+  allResponses: ResponsesByCode,
+  currentInputResponses: ListItem
+) => {
+  const currentResponses = Object.entries(currentInputResponses).reduce(
+    (responses, [code, value]) => ({ ...responses, [code]: value?.value }),
+    {}
+  )
+  return { ...allResponses, ...currentResponses }
+}
+
+export const buildElements = async (
+  fields: TemplateElement[],
+  allResponses: ResponsesByCode,
+  currentInputResponses: ListItem,
+  currentUser: User,
+  applicationData: ApplicationDetails
+) => {
+  const elements = fields.map((field, index) => ({
+    ...defaultEvaluatedElement,
+    id: index,
+    code: field.code,
+    pluginCode: field.elementTypePluginCode as string,
+    category: field.category as TemplateElementCategory,
+    title: field.title as string,
+    parameters: field.parameters,
+    validationExpression: field?.validation || true,
+    validationMessage: field?.validationMessage || '',
+    isVisibleExpression: field?.visibilityCondition || true,
+    isEditableExpression: field?.isEditable || true,
+    isRequiredExpression: field?.isRequired || true,
```

Comment:
Nice! This was broken on the current ListBuilder. Will check how it is using this version :)

Reply:
> Also required fields are showing for all elements in the Add ingredient modal, should only be showing for the ones with `isRequired` = true

Yeah -- once again I forget that that "false" will fallback to the second parameter of an `||` operator 🙄 -- have updated to `??`: 9835533f